### PR TITLE
docs(megatron): add GLM 5.2 two-node training proof

### PR DESCRIPTION
**Pull request by hershg**
## Summary

- Add the smallest standalone Megatron GLM 5.2 proof: two B300 nodes, rank-32 LoRA, 4 prompts × 4 rollouts, and two optimizer steps.
- Sample again after the first update and checkpoint every step.
- Keep runtime fixes and dependency changes in their focused upstream PRs; this branch adds only the example.

The original standalone run `glm52-m1a-native-dsa-recompute-20260829k` completed 32/32 rollouts, two forward/backward passes, two finite optimizer updates, two publications, and two durable checkpoints. An exact-head replay on current SkyRL main is required before this draft is ready to merge.

Native BF16 publication moved to #2117. Packed-result ordering remains #2043, Hadamard packaging remains #2095, and forwarding reliability remains #2097/#2100.

## Testing

```bash
bash -n examples/train/megatron/run_megatron_grpo_glm5_2_744b_lora.sh
bash format.sh
```

Shell syntax, Ruff, Black, and secret scanning pass.

### CODE REVIEW

**Comment by hershg** [comment]
@erictang000 early review welcome before Monday on the standalone Megatron example shape and module mapping. The draft now uses the smallest post-update fidelity proof: two disaggregated 8×B300 nodes, 4 prompts × 4 rollouts × 2 steps, with attention + shared-MLP + routed-expert LoRA active in vLLM. It stays draft until #2043, #2095, native MCore/Bridge pins, and the corrected live rerun are complete; no merge is requested yet.

**Comment by hershg** [comment]
Corrected live rerun is now active as `glm52-m1a-2step-20260828a` / service `skyrl-glm52-2n-direct-0-193d652d` on 2×8 B300. It uses SkyRL `ee28351f`, image digest `d3efc4bc…4edd`, full attention/shared-MLP/routed-expert LoRA targets, `4 prompts × 4 rollouts × 2 steps`, eval disabled, and checkpoint every step. Both pods are Running/claimed with zero restarts and are in the pinned CUDA 13/FHT bootstrap. Keeping this draft until the second post-update sampling round and native dependency pins are terminal.

**Comment by hershg** [comment]
Exact one-node B300 gate passed for this PR head 0d9238a21bdac46a3761adbdc592c4b53c9825a4. Driver 610.43.02, CUDA 13.0, PyTorch 2.11.0+cu130, TE 2.16.0, FlashAttention 2.8.3, vLLM 0.26.0, Bridge 37ee6c41, MCore 67b6ae70e, and FHT e7706faf were exercised. The full GLM MLAModelProvider constructed, an sm_103 extension compiled and ran, BF16 FHT reconstruction/gradient passed, and a real vLLM engine returned /health=200; no CUDA 803, zero restarts, exit 0. Run glm52-native-registry-preflight-20260829e. The unchanged two-node standalone 4x4x2 optimizer/checkpoint proof is now live.

File: examples/train/megatron/run_megatron_grpo_glm5_2_744b_lora.sh [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+# The defaults assume a shared home directory.
+
+MODEL_NAME="${MODEL_NAME:-zai-org/GLM-5.2-FP8}"
+DATA_DIR="${DATA_DIR:-$HOME/data/gsm8k}"
+CKPT_DIR="${CKPT_DIR:-$HOME/ckpts/glm5_2_744b_grpo_megatron}"
+LORA_SYNC_DIR="${LORA_SYNC_DIR:-$CKPT_DIR/lora-sync}"
+LOGGER="${LOGGER:-console}"
+
+TRAIN_FILE="${DATA_DIR}/train.parquet"
+test -f "${TRAIN_FILE}"
```
[the remark this answers is not in the document]
Comment (hershg) [reply examples/train/megatron/run_megatron_grpo_glm5_2_744b_lora.sh:23]: Fixed in a895b92. The example now prints the missing path and the exact GSM8K preparation command before exiting nonzero. Shell syntax and code-quality CI pass.

File: skyrl/backends/skyrl_train/distributed/megatron/megatron_strategy.py [hunk 2]
```diff
@@ -379,17 +378,16 @@ def save_checkpoint(
         if not async_save:
             # Async path keeps the pending request alive in the queue until its finalize;
             # tearing it down here would orphan that write.
-            ckpt_base.async_calls.close()
-            ckpt_base.async_calls = AsyncCallsQueue(persistent=True)
+            self._async_calls.close()
+            self._async_calls = AsyncCallsQueue(persistent=True)
```
[the remark this answers is not in the document]
Comment (hershg) [reply skyrl/backends/skyrl_train/distributed/megatron/megatron_strategy.py]: Fixed in a895b92. save_checkpoint now finalizes any preceding async request before every new save, including a synchronous or cloud save, before the queue can be closed and recreated. test_sync_save_waits_for_prior_async_save locks the finalize to sync save to close order. CI passes 1,532 train tests without vLLM and 193 with vLLM.

File: pyproject.toml [hunk 3]
```diff
@@ -347,15 +350,17 @@ torchvision = [
     { index = "pytorch-cpu", marker = "sys_platform == 'darwin'" },
 ]
 harbor = { git = "https://github.com/laude-institute/harbor", rev = "3de07a0e01f3368921766437fc7afece3ddec23d" }
-megatron-bridge = {git = "https://github.com/NVIDIA-NeMo/Megatron-Bridge", rev = "91a15142a4b4442a8d46ab539d1b923bd08570d0", marker = "sys_platform == 'linux'"}
-# megatron-core main branch: https://github.com/NVIDIA/Megatron-LM/tree/main latest as of 6/8/26
-megatron-core = {git = "https://github.com/NVIDIA/Megatron-LM", rev = "71e418ea7d7b3a6c9a53238c543c3e0b43e11026", marker = "sys_platform == 'linux'"}
+megatron-bridge = {git = "https://github.com/NVIDIA-NeMo/Megatron-Bridge", rev = "37ee6c41f02a10c59ddd345fc3dca1d645f581e3", marker = "sys_platform == 'linux'"}
+# Match Bridge's MCore base plus the focused FA4 and DSA recompute fixes in
+# NVIDIA/Megatron-LM#6964 and NVIDIA/Megatron-LM#6793.
+megatron-core = {git = "https://github.com/hershg/Megatron-LM", rev = "16512be1ac1ee440c8b869db4ad9ad5eafc1a056", marker = "sys_platform == 'linux'"}
```
[the remark this answers is not in the document]
Comment (hershg) [reply pyproject.toml]: Agreed; this is explicitly temporary. a895b92 documents in pyproject.toml that the personal integration pin must be replaced by an official NVIDIA commit once both upstream fixes land. The exact constituents are NVIDIA/Megatron-LM#6964 and #6793, and the PR body records both SHAs so this is immutable rather than a mutable fork reference.

File: skyrl/backends/skyrl_train/workers/worker_utils.py [hunk 4]
```diff
@@ -424,6 +424,18 @@ def reorder_and_combine_batches(self, batches: List[TensorBatch]) -> TensorBatch
...
+        """Restore per-sample microbatch outputs to input order."""
+        ordered = [None] * self.data.batch_size
+        for original_indices, items in zip(self._microbatches, batches):
+            if len(items) < len(original_indices):
+                raise ValueError("Microbatch output has fewer items than input samples")
+            for original_idx, item in zip(original_indices, items):
+                ordered[original_idx] = item
+        if any(item is None for item in ordered):
+            raise ValueError("Microbatch outputs do not cover every input sample")
+        return ordered
```
[the remark this answers is not in the document]
Comment (hershg) [reply skyrl/backends/skyrl_train/workers/worker_utils.py]: Fixed in a895b92. reorder_and_combine_items now fails immediately when fewer microbatch outputs are supplied, with a focused regression. CI passes 1,532 train tests without vLLM and 193 with vLLM.

**Comment by hershg** [comment]
@erictang000 Final head `ede580a` is ready for maintainer review. The two-node Megatron proof completed 32/32 rollouts, two optimizer steps, two durable checkpoints, and bounded sampler/trainer agreement; all review threads and CPU/train/code-quality checks are clean. GPU CI is fork-secret limited, with the exact B300 receipt pinned in the description. The current head is also live in cold TCLI qualification XID `1049067`; I will append its terminal receipt without changing the code.
